semgrep-core/src/matching/Generic_vs_generic.ml

```diff
@@ -1363,9 +1363,20 @@ and type_of_expr e : (G.ident * G.type_ option) option =
        * _params and _args to calculate the resulting type.
        *)
       | Some { t = TyFun (_params, tret); _ } -> Some (idb, Some tret)
-      | Some _ -> None
-      | None -> None)
-  | B.ParenExpr (_, e, _) -> type_of_expr e
+      | Some _
+      | None ->
+          None)
+  (* deep: in Java, there can be an implicit `this.`
+     so calculate the type in the same way as above
+     THINK: should we do this for all languages? Why not? *)
+  | B.Call ({ e = N (Id (idb, { B.id_type = tb; _ })); _ }, _args)
+    when lang = Lang.Java -> (
+      match !tb with
+      | Some { t = TyFun (_params, tret); _ } -> Some (idb, Some tret)
```

Comment:
But here we need to watch out that the function is fully applied perhaps, it may be a partial application. I guess this is not a problem in Java however.